helpers/metrics.py

```diff
@@ -0,0 +1,172 @@
+
+    # we add a id column to the labels dataset, which should not exist in detections too;
+    # this allows us to distinguish matching from non-matching detections
+    _labels_gdf['label_id'] = _labels_gdf.index
+    _dets_gdf['det_id'] = _dets_gdf.index
+    # We need to keep both geometries after sjoin to check the best intersection
+    _labels_gdf['label_geom'] = _labels_gdf.geometry
+    
+    # TRUE POSITIVES
+    left_join = gpd.sjoin(_dets_gdf, _labels_gdf, how='left', predicate='intersects', lsuffix='left', rsuffix='right')
+    
+    # Test that something is detected
+    candidates_tp_gdf = left_join[left_join.label_id.notnull()].copy()
+
+    # IoU computation between labels and detections
+    geom1 = candidates_tp_gdf['geometry'].to_numpy().tolist()
+    geom2 = candidates_tp_gdf['label_geom'].to_numpy().tolist()
+    iou = []
+    for (i, ii) in zip(geom1, geom2):
+        iou.append(intersection_over_union(i, ii))
+    candidates_tp_gdf['IOU'] = iou
+    
+    # Filter detections based on IoU value
+    best_matches_gdf = candidates_tp_gdf.groupby(['det_id'], group_keys=False).apply(lambda g:g[g.IOU==g.IOU.max()])
+    best_matches_gdf.drop_duplicates(subset=['det_id'], inplace=True) # <- this line could change the results depending and which is dropped 
+
+    # Detection, resp labels, with IOU lower than threshold value are considered as FP, resp FN, and saved as such
+    actual_matches_gdf = best_matches_gdf[best_matches_gdf['IOU'] >= iou_threshold].copy()
+    actual_matches_gdf = actual_matches_gdf.sort_values(by=['IOU'], ascending=False).drop_duplicates(subset=['label_id', 'tile_id'])
+
+    matched_det_ids = actual_matches_gdf['det_id'].unique().tolist()
+    matched_label_ids = actual_matches_gdf['label_id'].unique().tolist()
+    fp_gdf_temp = candidates_tp_gdf[~candidates_tp_gdf.det_id.isin(matched_det_ids)].drop_duplicates(subset=['det_id'], ignore_index=True)
+    fn_gdf_temp = candidates_tp_gdf[~candidates_tp_gdf.label_id.isin(matched_label_ids)].drop_duplicates(subset=['label_id'], ignore_index=True)
+    fn_gdf_temp.loc[:, 'geometry'] = fn_gdf_temp.label_geom
+
+    actual_matches_gdf['IOU'] = actual_matches_gdf.IOU.round(3)
+
+    # Test that it has the right class (id starting at 1 for labels and at 0 for detections)
+    condition = actual_matches_gdf.label_class == actual_matches_gdf.det_class+1
+    tp_gdf = actual_matches_gdf[condition].reset_index(drop=True)
+    mismatched_classes_gdf = actual_matches_gdf[~condition].reset_index(drop=True)
+    mismatched_classes_gdf.drop(columns=['x', 'y', 'z', 'dataset_right', 'label_geom'], errors='ignore', inplace=True)
+    mismatched_classes_gdf.rename(columns={'dataset_left': 'dataset'}, inplace=True)
+
+
+    # FALSE POSITIVES
+    fp_gdf = left_join[left_join.label_id.isna()].copy()
+    assert(len(fp_gdf[fp_gdf.duplicated()]) == 0)
+    fp_gdf = pd.concat([fp_gdf_temp, fp_gdf], ignore_index=True)
```

Comment:
Should we not check if there are duplicates?

Reply:
There are no duplicates since `fp_gdf = left_join[left_join.label_id.isna()].copy()` and `fp_gdf_temp` is deduced from `candidates_tp_gdf = left_join[left_join.label_id.notnull()].copy()`.